### ISSUE TITLE
On armv6/7, mc_cycle_counter () fails with SIGILL, Illegal instruction

### ISSUE THREAD

**Issue by adams-1979** [mirage/mirage-crypto#113]
I recently compiled and ran the following program on an Odroid U2 (ARM Cortex A9 Quad-Core) and also on a Raspberry Pi 1:
```
let () = Mirage_crypto_rng_unix.initialize ()
```
The result was a `Illegal instruction (core dumped)`.
The cause seems to be this line
```
__asm__ __volatile__("mrrc p15, 1, %0, %1, c14":"=r"(c_lo), "=r"(c_hi));
```
in `entropy_cpu_stubs.c`.
This tries to access the CNTVCT, Virtual Count register on the ARM CPU, presumably in order to seed the RNG.
However, according to the Documentation for the [ARM Architecture](https://developer.arm.com/documentation/ddi0406/c/System-Level-Architecture/System-Control-Registers-in-a-VMSA-implementation/VMSA-System-control-registers-descriptions--in-register-order/CNTVCT--Virtual-Count-register--VMSA?lang=en), these Generic Timer Registers are not accessible from user space. So this approach doesn't work and will fail on all ARMv7 and up.
To use the Timer Registers, you have to enable access via CNTKCTL from *kernel mode*, which basically means you have to write a kernel module.

After some searching I found this: [https://github.com/thoughtpolice/enable_arm_pmu](https://github.com/thoughtpolice/enable_arm_pmu) which uses a similar approach with the ARM's PMU registers, which are subject to the same restrictions as the generic timers. The author also has an interesting [Blog Post](http://neocontra.blogspot.com/2013/05/user-mode-performance-counters-for.html) explaining the matter in detail.

Cheers.

**Comment by hannesm** [comment]
Thanks for your report, and the references how a potential fix could look like. Since I don't have access to such hardware, and CI does not (yet?) support these architectures, I'm not really sure how I could add code and test this code. If you happen to have a patch for mirage-crypto that works on your hardware, I'm happy to review it.

**Comment by adams-1979** [comment]
I've spent some thought on this issue and I don't think there is a good, practical solution. The problem seems to arise only if the following two conditions are met:
1.) compiled for ARM
2.) linked to a program that runs in user space
For building unikernels, which I believe is the main use case of this library, this problem might not occur, even on ARM hardware. Shipping a kernel module with the library probably doesn't make much sense. Maybe a comment could be added somewhere near `mc_cycle_counter ()`, something like `/* Attention: this might fail when executed in non-privileged mode */`.
This would have saved me 1 hour with my favourite debugger (I'm not a technical guy, I do mostly HTML and Javascript stuff).
As far as I'm concerned, this issue can be closed.

**Comment by hannesm** [comment]
@adams-1979 thanks for your comment. The goal of `mirage-crypto-rng` is not limited to `MirageOS` (i.e. unikernels). So, I'd suggest a third thing:
If we're on ARM(7) (how can we detect this in `Mirage_crypto_rng_unix.initialize` / `Mirage_crypto_rng_lwt.initialize` -- or have the `read_virtual_count` use the low-resolution timers?) _and_ not in kernel space (but have a host operating system), do not collect entropy from "timer" (or use a low-resolution timer). The point is: entropy is fed via the `getrandom` / `getentropy` system call anyways, so it is fine to use the RNG even without data from the timer.

If you're interested in implementing this, please go ahead. I'd need to read up on available instructions, and also get my hands on actual hardware to test the code I'd write...

**Comment by adams-1979** [comment]
I have rewritten `read_virtual_count` like this:
```
static inline uint32_t read_virtual_count () {
  uint32_t c_lo, c_hi;
  __asm__ __volatile__("msr %0, cpsr":"=r"(c_lo));
  if (c_lo & 0x0f)
      __asm__ __volatile__("mrrc p15, 1, %0, %1, c14":"=r"(c_lo), "=r"(c_hi));
  else
      c_lo = 0;
  return c_lo;
}
```
This makes the `SIGILL` go away on my device. However, this returns 0 on *all* ARM systems if in mode PL0. If this has no impact on the cryptographic properties of the library, as you say, then I should be ok. I'm not sure if this is a good solution, though.

**Comment by hannesm** [comment]
@adams-1979 thanks for your investigation and proposed patch. The "returns 0 on all ARM systems if in mode PL0" worries me a bit -- my current assumption is that the SIGILL is only present on some ARM devices (though I'm no longer sure on which / which not -- the CI installed in this repository is certainly fine with the aarch64 code)...

Eventually @Timada can shed some light what to do here https://github.com/mirage/mirage-crypto/blob/main/src/native/entropy_cpu_stubs.c#L25-L44

Now, re-reading the code, I guess your patch is fine: it only modifies the `__arm__` case, not the `__aarch64__` case, correct?

**Comment by adams-1979** [comment]
From the documentation I've read, `__arm__` and `__aarch64__` seem to be mutually exclusive. `__aarch64__` is present on ARMs with the 64-bit extension (armv8-a onwards). So yes, the `__aarch64__` case is not affected.

**Comment by TImada** [comment]
Unfortunately, there seems to be no perfect solution for this issue ...

This problem happens as a MirageOS application runs as a user process if it is compiled with the `-t unix` option.
On the other hand, MirageOS running on top of a virtualized environment can have access to privileged registers as it is launched with higher privilege level than PL0.

MirageOS running on aarch32 with server virtualization seems deprecated, because "MirageOS with Xen + Mini-OS is not recommended now" and "MirageOS with Xen + Solo5 does not support aarch32".  
So single `__arm__` use would be enough now.

We will have to have a combination macro like `#if defined(__arm__) && defined(__VIRT__)` to distinguish if a target platform uses server virtualization or not, when we enable MirageOS running on aarch32 with server virtualization.

**Comment by hannesm** [comment]
Thanks for your comments, I first tried @adams-1979 patch (but our arm32 CI failed). Then I went to https://chromium.googlesource.com/external/gperftools/+/master/src/base/cycleclock.h and looked into their implementation. And afterwards I figured the comment (in the entropy_cpu_stubs.c file) "Xen masks the performance counters").

Please have a look into #120 and whether that version of entropy_cpu_stubs.c works for you @adams-1979 (also @TImada if interested). The original code (which used to work on Mini-OS & Xen & arm32) is still around, guarded by an `#ifdef __ocaml_freestanding__`. If you can confirm that #120 works, I'll merge and release.

For your convenience, the diff is inline here:
```
diff --git a/src/native/entropy_cpu_stubs.c b/src/native/entropy_cpu_stubs.c
index 5cb5c0f..798da1f 100644
--- a/src/native/entropy_cpu_stubs.c
+++ b/src/native/entropy_cpu_stubs.c
@@ -23,13 +23,53 @@
 #endif /* __i386__ || __x86_64__ */
 
 #if defined (__arm__)
-/* Replace with `read_virtual_count` from MiniOS when that symbol
- * gets exported. */
-static inline uint32_t read_virtual_count () {
+#include <sys/time.h>
+/* see https://github.com/mirage/mirage-crypto/issues/113 and
+https://chromium.googlesource.com/external/gperftools/+/master/src/base/cycleclock.h
+ */
+/*
+ * XXX
+ * The ideal timing source on ARM are the performance counters, but these are
+ * presently masked by Xen.
+ * It would work like this:
+
+#if defined (__ARM_ARCH_7A__)
+  // Disable counter overflow interrupts.
+  __asm__ __volatile__ ("mcr p15, 0, %0, c9, c14, 2" :: "r"(0x8000000f));
+  // Program the PMU control register.
+  __asm__ __volatile__ ("mcr p15, 0, %0, c9, c12, 0" :: "r"(1 | 16));
+  // Enable all counters.
+  __asm__ __volatile__ ("mcr p15, 0, %0, c9, c12, 1" :: "r"(0x8000000f));
+
+  // Read:
+  unsigned int res;
+  __asm__ __volatile__ ("mrc p15, 0, %0, c9, c13, 0": "=r" (res));
+*/
+static inline uint32_t read_virtual_count ()
+{
+#ifdef __ocaml_freestanding__
   uint32_t c_lo, c_hi;
   __asm__ __volatile__("mrrc p15, 1, %0, %1, c14":"=r"(c_lo), "=r"(c_hi));
   return c_lo;
+#else
+  uint32_t pmccntr;
+  uint32_t pmuseren;
+  uint32_t pmcntenset;
+  // Read the user mode perf monitor counter access permissions.
+  __asm__ __volatile__ ("mrc p15, 0, %0, c9, c14, 0" : "=r" (pmuseren));
+  if (pmuseren & 1) {  // Allows reading perfmon counters for user mode code.
+    __asm__ __volatile__ ("mrc p15, 0, %0, c9, c12, 1" : "=r" (pmcntenset));
+    if (pmcntenset & 0x80000000ul) {  // Is it counting?
+      __asm__ __volatile__ ("mrc p15, 0, %0, c9, c13, 0" : "=r" (pmccntr));
+      // The counter is set up to count every 64th cycle
+      return pmccntr;
+    }
+  }
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  return tv.tv_usec;
 }
+#endif /* __ocaml_freestanding__ */
 #endif /* arm */
 
 #if defined (__aarch64__)
```

**Comment by hannesm** [comment]
Thanks @adams-1979 for your report, this has been fixed in #120 which is part of the upcoming 0.10.0 release. @TImada, as mentioned, there is a `#ifdef __VIRT__` - which is in our case `__ocaml_freestanding__`.
